src/cpp/tensorbase.cpp

```diff
@@ -42,130 +43,151 @@ TensorBase::TensorBase(const std::string& name,
     owned by the tensor.
     */
 
-    this->_check_inputs(data, name, dims);
-    this->_name = name;
-    this->_type = type;
-    this->_dims = dims;
+    _check_inputs(data, name, dims);
+    _name = name;
+    _type = type;
+    _dims = dims;
 }
 
+// TensorBase copy constructor
 TensorBase::TensorBase(const TensorBase& tb)
 {
-    /* This is the copy constructor for TensorBase.
-    A deep copy of the tensor data is performed here.
-    */
-    this->_dims = std::vector<size_t>(tb._dims);
-    this->_name = std::string(tb._name);
-    this->_type = TensorType(tb._type);
+    // check for self-assignment
+    if (&tb == this)
+        return;
+
+    // deep copy of tensor data
+    _dims = std::vector<size_t>(tb._dims);
+    _name = std::string(tb._name);
+    _type = TensorType(tb._type);
 }
 
+// TensorBase move constructor
 TensorBase::TensorBase(TensorBase&& tb)
 {
-    /* This is the move constructor for TensorBase.
-    */
-    this->_name = std::move(tb._name);
-    this->_type = std::move(tb._type);
-    this->_dims = std::move(tb._dims);
-    this->_data = tb._data;
-    tb._data = 0;
+    // check for self-assignment
+    if (&tb == this)
+        return;
+
+    // Move data
+    _name = std::move(tb._name);
+    _type = std::move(tb._type);
+    _dims = std::move(tb._dims);
+    _data = tb._data;
+
+    // Mark that the data is no longer owned by the source
+    tb._data = NULL;
 }
 
+// TensorBase destructor
 TensorBase::~TensorBase()
 {
-    if(this->_data)
-        free(this->_data);
+    if (_data != NULL) {
+        delete(reinterpret_cast<unsigned char *>(_data));
```

Comment:
I'm curious why the cast to ``unsigned char *``.  Doesn't ``delete`` take ``void*``?

Reply:
I think I was getting a warning for something. I'll pull the cast and recompile to confirm